notebooks/1.processing-profiles/nbconverted/1.preprocessing-features.py

```diff
@@ -0,0 +1,198 @@
+        left_on=["treatment", "well_position"],
+        right_on=["UCD ID", "Well"],
+        how="left",
+    )
+
+    # Save the augmented platemap with MOA information to the output directory
+    merged_df.to_csv(results_platemap_dir / outname, index=False)
+
+
+# In[5]:
+
+
+# Load the original barcode metadata
+barcodes = io_utils.load_barcodes(barcode_path=barcode_path)
+
+# Initialize a list to store the updated barcode entries
+updated_barcodes = []
+
+# Iterate through each batch and its associated plates in the barcode metadata
+for batch_id, plates in barcodes.items():
+    # Generate the updated platemap name with the new tag
+    updated_platemap_name = f"{list(plates.keys())[0]}_{moa_tag}"
+
+    # Iterate through all profiles associated with the current batch of plates
+    for plate_profiles in plates.values():
+        for profile_name in plate_profiles:
+            # Append the updated profile-barcode mapping
+            updated_barcodes.append([profile_name, updated_platemap_name])
+
+# Convert the updated barcode list into a DataFrame
+updated_barcodes_df = pd.DataFrame(
+    updated_barcodes, columns=["plate_barcode", "platemap_file"]
+)
+
+updated_barcodes_df.to_csv(updated_barcode_path, index=False)
+updated_barcodes_df
+
+
+# Next, we aim to **annotate the single-cell profiles** by integrating them with the newly augmented platemaps. This process begins by loading the updated barcode metadata, which now includes MOA information. Using these barcodes, batch profiles are loaded from the specified data directory. The batch profiles are then annotated with additional metadata by aligning the platemap's well positions (`Metadata_well_position`) with the profile metadata (`Image_Metadata_Well`). Finally, the annotated profiles are concatenated at the batch level, combining all plates within each batch into a single unified dataset.
+
+# In[6]:
+
+
+# Load the updated barcode metadata containing MOA information
+updated_barcodes = io_utils.load_barcodes(barcode_path=updated_barcode_path)
+
+# Load batch profiles using the updated barcodes
+batch_profiles = io_utils.batch_load_profiles(
+    path_to_data_dir=data_dir, barcodes=updated_barcodes
+)
+
+# Annotate the loaded batch profiles with metadata
+annotated_profiles = data_utils.annotate_batch(
+    batch_profiles,
+    metadata_dir=results_platemap_dir,
+    platemap_metadata_query="Metadata_well_position",
+    profile_metadata_query="Image_Metadata_Well",
+)
+
+# concatenate all plates in a batch into one
+concat_batch_profiles = data_utils.concat_batch_profiles(annotated_profiles)
+
+
+# Finally, we **aggregate the single-cell profiles** into two distinct sets:
+#
+# 1. **Replicate-Level Aggregation**: Profiles are combined at the replicate level, grouping data from individual experimental replicates.
+# 2. **MOA-Level Aggregation**: Profiles are further aggregated based on their shared **Mechanism of Action (MOA)**, enabling pathway-specific analysis.
+
```

Comment:
This is for all 4 plates combined, correct?

Reply:
Correct! These aggregate profiles are generated by combining all plates! (batch)